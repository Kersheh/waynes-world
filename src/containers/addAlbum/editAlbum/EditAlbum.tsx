import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useForm, FormProvider } from 'react-hook-form';
import { isEqual } from 'lodash';

import { RootState, EditAlbumFormValues } from 'types';
import TextInput from 'components/textInput/TextInput';
import TextArea from 'components/textArea/TextArea';
import Button from 'components/button/Button';
import IconCaret from 'components/icons/IconCaret';
import { saveUpdateAlbumAction, setEditAlbumAction } from '../addAlbumActions';
import styles from './EditAlbum.module.scss';

interface EditAlbumProps {
  setShowEditAlbum: React.Dispatch<React.SetStateAction<boolean>>;
}
const EditAlbum = ({ setShowEditAlbum }: EditAlbumProps) => {
  const dispatch = useDispatch();
  const { editAlbum, editAlbumId } = useSelector(
    (state: RootState) => state.addAlbum
  );

  const methods = useForm<EditAlbumFormValues>({
    defaultValues: editAlbum,
    mode: 'onChange'
  });

  // update store values on change
  const watchedValues = methods.watch();
  useEffect(() => {
    const data = methods.getValues();
    if (Object.keys(watchedValues).length > 0 && !isEqual(editAlbum, data)) {
      dispatch(
        setEditAlbumAction({
          album: data
        })
      );
    }
  }, [dispatch, methods, watchedValues, editAlbum]);

  return (
    <div className={styles.editAlbum}>
      <div className={styles.header}>
        <Button onClick={() => setShowEditAlbum(false)}>
          <IconCaret className={styles.left} />
          <span>Back</span>
        </Button>

        <span>{editAlbumId ? 'Edit Album' : 'Add New Album'}</span>

        <Button
          onClick={async () => {
            if (await methods.trigger()) {
              dispatch(
                saveUpdateAlbumAction({
                  id: editAlbumId,
                  album: methods.getValues()
                })
              );
            }
          }}
        >
          Save
          <IconCaret className={styles.right} />
        </Button>
      </div>

      <FormProvider {...methods}>
        <form>
          <div className={styles.artInfo}>
            <div className={styles.art}>
              <div className={styles.artText}>Album Art</div>
              <div className={styles.artImage} />
            </div>

            <div className={styles.info}>
              <TextInput
                name="artist"
                label="Artist"
                maxLength={50}
                required
                error={methods.errors?.artist?.message}
              />
              <TextInput
                name="album"
                label="Album"
                maxLength={50}
                required
                error={methods.errors?.album?.message}
              />
            </div>
          </div>

          <TextInput name="genre" label="Genre" maxLength={20} />

          <div className={styles.otherInfo}>
            <TextInput
              name="year"
              label="Year"
              maxLength={4}
              onChange={e => {
                methods.setValue('year', e.target.value.replace(/[^0-9]/i, ''));
              }}
            />
            <TextInput name="shelf" label="Shelf" maxLength={10} />
          </div>

          <TextArea name="comments" label="Comments" maxLength={2000} />
        </form>
      </FormProvider>
    </div>
  );
};

export default EditAlbum;
