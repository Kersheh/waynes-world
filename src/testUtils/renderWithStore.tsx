import React from 'react';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware, compose, Reducer } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { render } from '@testing-library/react';

import rootReducer from 'store/reducers';
import rootSaga from 'store/sagas';

const renderWithStore = (component: React.ReactElement) => {
  const sagaMiddleware = createSagaMiddleware();
  const store = createStore(
    rootReducer as Reducer,
    compose(applyMiddleware(sagaMiddleware))
  );
  sagaMiddleware.run(rootSaga);

  return render(<Provider store={store}>{component}</Provider>);
};

export default renderWithStore;
