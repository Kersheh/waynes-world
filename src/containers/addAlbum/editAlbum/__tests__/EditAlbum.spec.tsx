import React from 'react';

import renderWithStore from 'testUtils/renderWithStore';

import EditAlbum from '../EditAlbum';

describe('<EditAlbum/> component', () => {
  it('should render', () => {
    const { container } = renderWithStore(<EditAlbum setShowEditAlbum={() => {}} />);

    expect(container).toMatchSnapshot();
  });
});
