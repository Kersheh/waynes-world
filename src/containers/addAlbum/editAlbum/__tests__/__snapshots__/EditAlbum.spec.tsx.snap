// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditAlbum/> component should render 1`] = `
<div>
  <div
    class="editAlbum"
  >
    <div
      class="header"
    >
      <button
        class="button text"
        type="button"
      >
        <svg
          aria-labelledby="fqsi-ant-right-title"
          class="left"
          id="si-ant-right"
          viewBox="0 0 574 1024"
        >
          <path
            d="M10 9Q0 19 0 32t10 23l482 457L10 969Q0 979 0 992t10 23q10 9 24 9t24-9l506-480q10-10 10-23t-10-23L58 9Q48 0 34 0T10 9z"
          />
        </svg>
        <span>
          Back
        </span>
      </button>
      <span>
        Add New Album
      </span>
      <button
        class="button text"
        type="button"
      >
        Save
        <svg
          aria-labelledby="fqsi-ant-right-title"
          class="right"
          id="si-ant-right"
          viewBox="0 0 574 1024"
        >
          <path
            d="M10 9Q0 19 0 32t10 23l482 457L10 969Q0 979 0 992t10 23q10 9 24 9t24-9l506-480q10-10 10-23t-10-23L58 9Q48 0 34 0T10 9z"
          />
        </svg>
      </button>
    </div>
    <form>
      <div
        class="artInfo"
      >
        <div
          class="art"
        >
          <div
            class="artText"
          >
            Album Art
          </div>
          <div
            class="fileUpload"
          >
            <button
              class="button basic"
              type="button"
            >
              <input
                type="file"
              />
              <div
                class="artImage"
              />
              <div
                class="artPlus"
              >
                <svg
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M24 10 h-10 v-10 h-2 v10 h-10 v2 h10 v10 h2 v-10 h10 z"
                    fill="white"
                  />
                </svg>
              </div>
            </button>
          </div>
        </div>
        <div
          class="info"
        >
          <div
            class="textInput"
          >
            <div
              class="label"
            >
              Artist
               *
            </div>
            <input
              class=""
              maxlength="50"
              name="artist"
              required=""
              type="text"
            />
            <div
              class="counter"
            >
              0
               / 
              50
            </div>
          </div>
          <div
            class="textInput"
          >
            <div
              class="label"
            >
              Album
               *
            </div>
            <input
              class=""
              maxlength="50"
              name="album"
              required=""
              type="text"
            />
            <div
              class="counter"
            >
              0
               / 
              50
            </div>
          </div>
        </div>
      </div>
      <div
        class="textInput"
      >
        <div
          class="label"
        >
          Genre
        </div>
        <input
          class=""
          maxlength="20"
          name="genre"
          type="text"
        />
        <div
          class="counter"
        >
          0
           / 
          20
        </div>
      </div>
      <div
        class="otherInfo"
      >
        <div
          class="textInput"
        >
          <div
            class="label"
          >
            Year
          </div>
          <input
            class=""
            maxlength="4"
            name="year"
            type="text"
          />
          <div
            class="counter"
          >
            0
             / 
            4
          </div>
        </div>
        <div
          class="textInput"
        >
          <div
            class="label"
          >
            Shelf
          </div>
          <input
            class=""
            maxlength="10"
            name="shelf"
            type="text"
          />
          <div
            class="counter"
          >
            0
             / 
            10
          </div>
        </div>
      </div>
      <div
        class="textArea"
      >
        <div
          class="label"
        >
          Comments
        </div>
        <textarea
          maxlength="2000"
          name="comments"
        />
        <div
          class="counter"
        >
          0
           / 
          2000
        </div>
      </div>
    </form>
  </div>
</div>
`;
